File: ruby/flows/charge_test.rb
# frozen_string_literal: true

# Flow: create (mode=charge) → sign → charge, via rail0-ruby.
#
# Run:
#   set -a && source ../../.env && set +a
#   bundle exec ruby -Iflows flows/charge_test.rb

require_relative "test_helper"

describe "charge (rail0-ruby)" do
  it "charges a payment in one shot" do
    client   = new_client
    payee_pk = env("ACCOUNT_PRIVATE_KEY")
    pm       = discover_payment_method(client)

    puts "→ creating payment (mode=charge) and submitting payer signature"
    payment_id = create_and_sign(client, pm, "charge")
    puts "  payment_id=#{payment_id}"

    puts "→ charge/prepare"
    prep = payee_client.payments.charge_prepare(payment_id)
    payee_client.payments.charge(payment_id, { signed_transaction: sign_tx(prep[:unsigned_transaction], payee_pk) })
    final = poll_until_status(client, payment_id, "charged")
    puts "  charged — status=#{final[:status]}"
  end
end

File: cli/flows/charge_test.go
package flows_test

import "testing"

// charge (immediate pay-through, no escrow), driven through the rail0 CLI.
func TestCharge(t *testing.T) {
	payeeKey := env(t, "ACCOUNT_PRIVATE_KEY")
	desc(t, "charge (immediate pay-through, no escrow)",
		"1. Create payment + payer signature (charge mode)",
		"2. Charge — pay through to payee in one call (payee)")

	step(t, "1. create + sign (charge mode)")
	rail0Id := createSigned(t, "charge")

	step(t, "2. charge/prepare → sign → submit")
	runCLI(t, "payments", "charge", rail0Id, "-p", payeeKey)
	res := pollStatus(t, rail0Id, "charge", "charged")
	ok(t, "charged — status=%s", res["status"])
}
